### ISSUE TITLE
Relocate caffe set device/mode to generation method

### ISSUE THREAD

**Issue by Purg** [Kitware/SMQTK#328]
TL;DR: The if-statement at the top of ``_setup_network`` that sets the caffe device and mode should be moved to the head of the ``compute_descriptor_async`` method.

If the Caffe descriptor generator is used in a multi-threaded/process environment, the Caffe device/mode setting only affects the thread/process it was called on. Since the device/mode is set on construction via the ``_setup_network`` method, if a separate thread/process that shares/picks-up the algorithm and calls compute_descriptor, on the CPU will be used (default caffe mode).

Moving the caffe set device/mode calls to the ``compute_descriptor_async`` method would instead cause the deivce/mode to be set on the appropriate thread/process where descriptors are actually requested.

**Comment by mattdawkins** [comment]
I had this same problem in VIDTK
